### PR TITLE
En-6274 : add new data fields in elastic seach transaction

### PR DESCRIPTION
**Pull request by miiu96**

### CODE REVIEW

File: core/indexer/disabled/disabledTxLogsProcessorDatabase.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	return new(txLogProcessor)
+}
+
+// GetLogFromRAM -
+func (t *txLogProcessor) GetLogFromRAM(_ []byte) (data.LogHandler, bool) {
+	return nil, false
+}
+
+// SaveLogsAlsoInRAM -
+func (t *txLogProcessor) SaveLogsAlsoInRAM() {
```
Comment (sasurobert) [inline core/indexer/disabled/disabledTxLogsProcessorDatabase.go:24]: GetLogFromCache - SaveLogToCache

File: core/indexer/nilIndexer.go [hunk 2]
```diff
@@ -18,6 +19,11 @@ func NewNilIndexer() *NilIndexer {
 func (ni *NilIndexer) SaveBlock(_ data.BodyHandler, _ data.HeaderHandler, _ map[string]data.TransactionHandler, _ []uint64, _ []string) {
 }
 
+// SetTxLogsProcessor will do nothing
+func (ni *NilIndexer) SetTxLogsProcessor(_ process.TransactionLogProcessorDatabase) {
+
```
Comment (sasurobert) [inline core/indexer/nilIndexer.go:24]: delete empty line

File: go.mod [hunk 3]
```diff
@@ -31,6 +31,7 @@ require (
 	github.com/libp2p/go-libp2p-kad-dht v0.2.1
 	github.com/libp2p/go-libp2p-kbucket v0.2.1
 	github.com/libp2p/go-libp2p-pubsub v0.1.1
+	github.com/magiconair/properties v1.8.0
```
Comment (sasurobert) [inline go.mod:34]: why is this needed ?

File: node/mock/indexerMock.go [hunk 4]
```diff
@@ -17,6 +18,11 @@ func (im *IndexerMock) SaveBlock(_ data.BodyHandler, _ data.HeaderHandler, _ map
 	panic("implement me")
 }
 
+// SetTxLogsProcessor will do nothing
+func (im *IndexerMock) SetTxLogsProcessor(_ process.TransactionLogProcessorDatabase) {
+
```
Comment (sasurobert) [inline node/mock/indexerMock.go:23]: delete empty line

File: process/interface.go [hunk 5]
```diff
@@ -237,6 +237,13 @@ type TransactionLogProcessor interface {
 	IsInterfaceNil() bool
 }
 
+// TransactionLogProcessorDatabase is interface the  for saving logs also in RAM
+type TransactionLogProcessorDatabase interface {
+	GetLogFromRAM(txHash []byte) (data.LogHandler, bool)
+	SaveLogsAlsoInRAM()
+	RemoveLogsFromRAM()
+}
```
Comment (sasurobert) [inline process/interface.go:245]: add IsInterfaceNil . change RAM to Cache, saveLogsToCache

File: process/mock/indexerMock.go [hunk 6]
```diff
@@ -18,6 +19,11 @@ func (im *IndexerMock) SaveBlock(body data.BodyHandler, header data.HeaderHandle
 	}
 }
 
+// SetTxLogsProcessor will do nothing
+func (im *IndexerMock) SetTxLogsProcessor(_ process.TransactionLogProcessorDatabase) {
+
```
Comment (sasurobert) [inline process/mock/indexerMock.go:24]: delete empty line

File: process/transactionLog/process.go [hunk 7]
```diff
@@ -25,13 +26,16 @@ type ArgTxLogProcessor struct {
 }
 
 type txLogProcessor struct {
-	storer      storage.Storer
-	marshalizer marshal.Marshalizer
+	storeLogsInRAM bool
```
Comment (sasurobert) [inline process/transactionLog/process.go:29]: change RAM to cache

File: process/transactionLog/process.go [hunk 8]
```diff
@@ -61,6 +68,30 @@ func (tlp *txLogProcessor) GetLog(txHash []byte) (data.LogHandler, error) {
...
+	return nil, false
+}
+
+// SaveLogsAlsoInRAM set will a flag with true and txLogProcessor will start save transactions logs also in RAM
+func (tlp *txLogProcessor) SaveLogsAlsoInRAM() {
+	tlp.storeLogsInRAM = true
+}
+
+// RemoveLogsFromRAM will remove all transaction logs from RAM memory
+func (tlp *txLogProcessor) RemoveLogsFromRAM() {
```
Comment (sasurobert) [inline process/transactionLog/process.go:89]: rename to Clean()

File: process/transactionLog/process.go [hunk 9]
```diff
@@ -93,6 +124,12 @@ func (tlp *txLogProcessor) SaveLog(txHash []byte, tx data.TransactionHandler, lo
 		})
 	}
 
+	if tlp.storeLogsInRAM {
+		tlp.mut.Lock()
```
Comment (sasurobert) [inline process/transactionLog/process.go:128]: make a small function with this if and save.

File: core/indexer/disabled/disabledTxLogsProcessorDatabase.go (hunk 1 above)
Comment (miiu96) [reply core/indexer/disabled/disabledTxLogsProcessorDatabase.go:24]: done.

File: core/indexer/nilIndexer.go (hunk 2 above)
Comment (miiu96) [reply core/indexer/nilIndexer.go:24]: done.

File: go.mod (hunk 3 above)
Comment (miiu96) [reply go.mod:34]: is not needed

File: node/mock/indexerMock.go (hunk 4 above)
Comment (miiu96) [reply node/mock/indexerMock.go:23]: done.

File: process/interface.go (hunk 5 above)
Comment (miiu96) [reply process/interface.go:245]: done.

File: process/mock/indexerMock.go (hunk 6 above)
Comment (miiu96) [reply process/mock/indexerMock.go:24]: done.

File: process/transactionLog/process.go (hunk 7 above)
Comment (miiu96) [reply process/transactionLog/process.go:29]: done.

File: process/transactionLog/process.go (hunk 8 above)
Comment (miiu96) [reply process/transactionLog/process.go:89]: done.

File: process/transactionLog/process.go (hunk 9 above)
Comment (miiu96) [reply process/transactionLog/process.go:128]: done.

File: process/transactionLog/process.go [hunk 10]
```diff
@@ -61,6 +68,30 @@ func (tlp *txLogProcessor) GetLog(txHash []byte) (data.LogHandler, error) {
...
+
+	if txLog, ok := tlp.logs[string(txHash)]; ok {
+		return txLog, true
+	}
+
+	return nil, false
+}
+
+// SaveLogToCache set will a flag with true and txLogProcessor will start save transactions logs also in RAM
+func (tlp *txLogProcessor) SaveLogToCache() {
```
Comment (ccorcoveanu) [inline process/transactionLog/process.go:84]: I think this should be renamed, the name suggests that it is actually saving the logs, instead it just sets the flag for enabling this functionality

File: core/indexer/processTransactions.go [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+	}
+
+	for hash, nrScResult := range countScResults {
+		if nrScResult < 2 {
+			transactions[hash].Status = txStatusNotExecuted
+		}
+	}
+
+	for hash, tx := range transactions {
+		log, ok := tdp.txLogsProcessor.GetLogFromCache([]byte(hash))
```
Comment (ccorcoveanu) [inline core/indexer/processTransactions.go:86]: `log` variable shadows another declaration

File: core/indexer/elasticsearchDatabase.go [hunk 12]
```diff
@@ -223,6 +226,12 @@ func (esd *elasticSearchDatabase) SaveTransactions(
 	}
 }
 
+// SetTxLogsProcessor will set tx logs processor
+func (esd *elasticSearchDatabase) SetTxLogsProcessor(txLogsProc process.TransactionLogProcessorDatabase) {
+	esd.txLogsProcessor = txLogsProc
+	esd.txLogsProcessor.SaveLogToCache()
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:232]: done. moved esd.txLogsProcessor.SaveLogToCache() from there

File: core/indexer/data.go [hunk 13]
```diff
@@ -9,22 +9,54 @@ import (
...
+	Sender               string        `json:"sender"`
+	ReceiverShard        uint32        `json:"receiverShard"`
+	SenderShard          uint32        `json:"senderShard"`
+	GasPrice             uint64        `json:"gasPrice"`
+	GasLimit             uint64        `json:"gasLimit"`
+	Data                 string        `json:"data"`
+	Signature            string        `json:"signature"`
+	Timestamp            time.Duration `json:"timestamp"`
+	Status               string        `json:"status"`
+	ReceiptValue         string        `json:"gasRefund"`
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/data.go]: Refactored now is `GasUsed`

File: core/indexer/processTransactions.go [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+		tx, ok := transactions[string(scResult.OriginalTxHash)]
+		if !ok {
+			continue
+		}
+		tx.SmartContractResults = append(tx.SmartContractResults, tdp.commonProcessor.convertScResultInDatabaseScr(scResult))
+		countScResults[string(scResult.OriginalTxHash)]++
+	}
+
+	for hash, nrScResult := range countScResults {
+		if nrScResult < 2 {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/processTransactions.go:96]: done.

File: core/indexer/processTransactions.go (hunk 11 above)
Comment (miiu96) [reply core/indexer/processTransactions.go:102]: done.

File: core/indexer/elasticsearchDatabase.go [hunk 15]
```diff
@@ -231,46 +240,16 @@ func (esd *elasticSearchDatabase) buildTransactionBulks(
...
-		if mb.Type == block.PeerBlock {
-			continue
-		}
-
-		mbMarshal, err := esd.marshalizer.Marshal(mb)
-		if err != nil {
-			log.Debug("indexer: marshal", "error", "could not marshal miniblock")
+	bulks := make([][]*Transaction, (len(txs)/txBulkSize)+1)
+	for i := 0; i < len(bulks); i++ {
+		if i == len(bulks)-1 {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/elasticsearchDatabase.go:247]: bulks[i] = append(bulks[i], txs[i*txBulkSize:]...) this is working fine without make([]*Transaction, txBulkSize)

File: process/transactionLog/process.go (hunk 10 above)
Comment (miiu96) [reply process/transactionLog/process.go:84]: done./

File: core/indexer/common.go [hunk 16]
```diff
@@ -187,6 +187,17 @@ func (cm *commonProcessor) buildRewardTransaction(
 }
 
 func (cm *commonProcessor) convertScResultInDatabaseScr(sc *smartContractResult.SmartContractResult) ScResult {
+	encodedData := strings.Split(string(sc.Data), "@")
+
+	decodedData := ""
+	for _, enc := range encodedData {
+		val, err := hex.DecodeString(enc)
+		if err != nil || len(val) == 0 {
+			continue
```
Comment (sasurobert) [inline core/indexer/common.go:196]: add enc to the list if it is not possible to decode

Comment (miiu96) [reply core/indexer/common.go]: done

File: core/indexer/common.go [hunk 17]
```diff
@@ -242,32 +191,57 @@ func (cm *commonProcessor) buildRewardTransaction(
...
+}
+
+func decodeScResultData(scrData []byte) string {
+	encodedData := strings.Split(string(scrData), "@")
+	encodedData = append([]string(nil), encodedData[1:]...)
+
+	decodedData := ""
+	for i, enc := range encodedData {
+		if !canInterpretAsString([]byte(enc)) {
+			continue
```
Comment (sasurobert) [inline core/indexer/common.go:219]: here decodedData should get the actual encoded data.

File: process/transactionLog/process.go [hunk 18]
```diff
@@ -61,6 +68,30 @@ func (tlp *txLogProcessor) GetLog(txHash []byte) (data.LogHandler, error) {
...
+	defer tlp.mut.RUnlock()
+
+	if txLog, ok := tlp.logs[string(txHash)]; ok {
+		return txLog, true
+	}
+
+	return nil, false
+}
+
+// EnableLogToBeSavedInCache will a flag with true and txLogProcessor will start save transactions logs also in RAM
```
Comment (ccorcoveanu) [inline process/transactionLog/process.go:83]: "will **set** a flag", "will start **saving**"

Comment (miiu96) [reply process/transactionLog/process.go:83]: done.
